File: src/tir/ir/layout/layout.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "utils.h"

namespace tvm {
namespace tir {

/**************** TLayout ****************/
ffi::Map<ffi::String, PrimExpr> TLayoutNode::Apply(const ffi::Array<PrimExpr>& coord,
                                                   const ffi::Array<PrimExpr>& shape) const {
  TVM_FFI_ICHECK_EQ(coord.size(), shape.size())
      << "ValueError: The size of coord and shape should be equal";
  return Apply(FlattenCoord(coord, shape));
}

TVM_FFI_STATIC_INIT_BLOCK() {
  namespace refl = tvm::ffi::reflection;
  auto def = refl::GlobalDef();
  def.def("tir.TLayoutCompatibleWithShape",
          [](TLayout layout, Array<PrimExpr> shape) { return layout->CompatibleWithShape(shape); });
  def.def("tir.TLayoutVerifyWellFormed", [](TLayout layout) { return layout->VerifyWellFormed(); });
  def.def("tir.TLayoutGetSize", [](TLayout layout, ffi::Optional<ffi::String> axis_name) {
    return layout->GetSize(axis_name);
  });
  def.def("tir.TLayoutGetSpan", [](TLayout layout, ffi::Optional<ffi::String> axis_name) {
    return layout->GetSpan(axis_name);
  });
  def.def("tir.TLayoutApplyWithShape",
          [](TLayout layout, ffi::Array<PrimExpr> coord, ffi::Array<PrimExpr> shape) {
            return layout->Apply(coord, shape);
          });
  def.def("tir.TLayoutApply",
          [](TLayout layout, ffi::Array<PrimExpr> coord) { return layout->Apply(coord); });
  def.def("tir.TLayoutApplyLinear",
          [](TLayout layout, PrimExpr coord) { return layout->Apply(coord); });
  def.def("tir.TLayoutCanonicalize", [](TLayout layout) { return layout->Canonicalize(); });
  def.def("tir.TLayoutTile", [](TLayout layout, TileLayout outer, ffi::Array<PrimExpr> outer_shape,
                                ffi::Array<PrimExpr> inner_shape) {
    return layout->Tile(outer, outer_shape, inner_shape);
  });
  def.def("tir.TLayoutDirectSum",
          [](TLayout layout, TileLayout left, ffi::Array<PrimExpr> left_shape,
             ffi::Array<PrimExpr> right_shape) {
            return layout->DirectSum(left, left_shape, right_shape);
          });
  def.def("tir.TLayoutIsTileInner",
          [](TLayout layout, TLayout tile_layout, ffi::Array<PrimExpr> tiled_shape,
             ffi::Array<PrimExpr> inner_shape) {
            return layout->IsTileInner(tile_layout, tiled_shape, inner_shape);
          });
  def.def("tir.TLayoutIsTileOuter",
          [](TLayout layout, TLayout tile_layout, ffi::Array<PrimExpr> tiled_shape,
             ffi::Array<PrimExpr> outer_shape) {
            return layout->IsTileOuter(tile_layout, tiled_shape, outer_shape);
          });
  def.def("tir.TLayoutIsDirectSumRight",
          [](TLayout layout, TLayout sum_layout, ffi::Array<PrimExpr> interleaved_shape,
             ffi::Array<PrimExpr> right_shape) {
            return layout->IsDirectSumRight(sum_layout, interleaved_shape, right_shape);
          });
  def.def("tir.TLayoutIsDirectSumLeft",
          [](TLayout layout, TLayout sum_layout, ffi::Array<PrimExpr> interleaved_shape,
             ffi::Array<PrimExpr> left_shape) {
            return layout->IsDirectSumLeft(sum_layout, interleaved_shape, left_shape);
          });
  def.def("tir.TLayoutSlice",
          [](TLayout layout, ffi::Array<PrimExpr> shape, Region region) -> ffi::Optional<TLayout> {
            return layout->Slice(shape, region);
          });
}

}  // namespace tir
}  // namespace tvm

File: rust/tirx-ext/rust/src/layout.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

//! Layout self-check: every Rust field offset must equal the C++ side's
//! reflected field offset — the ground-truth proof that the hand-mirrored
//! `#[repr(C)]` structs in node.rs are byte-compatible with the C++ nodes.
//! Exposed as the `tirx_ext._check_layouts` global and run on every Python
//! import, so a tirx header change that shifts a field fails loudly at load
//! time instead of reading the wrong pointer slot at count time.

use std::mem::offset_of;
use std::ops::ControlFlow;

use tvm_ffi::tvm_ffi_sys::{TVMFFIByteArray, TVMFFITypeKeyToIndex};

use crate::node::{
    AddNode, ExprNode, ForNode, IfThenElseNode, IntImmNode, VarNode, WhileNode,
};
use crate::reflect::for_each_field;

/// The reflected absolute byte offset of `field` in the C++ type `type_key`,
/// searching the type and then its ancestors (single-inheritance chain).
fn reflect_offset(type_key: &str, field: &str) -> Result<i64, String> {
    unsafe {
        let key = TVMFFIByteArray::from_str(type_key);
        let mut idx: i32 = 0;
        if TVMFFITypeKeyToIndex(&key, &mut idx) != 0 {
            return Err(format!("type key `{type_key}` is not registered"));
        }
        for_each_field(idx, |f| {
            if f.name.as_str() == field {
                ControlFlow::Break(f.offset)
            } else {
                ControlFlow::Continue(())
            }
        })
        .ok_or_else(|| format!("field `{field}` not found in `{type_key}` or its ancestors"))
    }
}

/// Verify every field of every node this crate mirrors. Returns the full list
/// of mismatches (empty = layouts proven identical).
pub(crate) fn check_layouts() -> Result<(), String> {
    // (rust offset, C++ type key, C++ reflected field name, label)
    macro_rules! entry {
        ($node:ty, $key:literal, $field:ident) => {
            entry!($node, $key, $field, stringify!($field))
        };
        ($node:ty, $key:literal, $field:ident, $cxx:expr) => {
            (
                offset_of!($node, $field) as i64,
                $key,
                $cxx,
                concat!(stringify!($node), ".", stringify!($field)),
            )
        };
    }
    let table = [
        // Expr base: span @24, ty @32 (the reflected `ir.Expr`).
        entry!(ExprNode, "ir.Expr", span),
        entry!(ExprNode, "ir.Expr", ty),
        entry!(VarNode, "tirx.Var", name_hint, "name"),
        entry!(IntImmNode, "ir.IntImm", value),
        entry!(AddNode, "tirx.Add", a),
        entry!(AddNode, "tirx.Add", b),
        entry!(ForNode, "tirx.For", loop_var),
        entry!(ForNode, "tirx.For", min),
        entry!(ForNode, "tirx.For", extent),
        entry!(ForNode, "tirx.For", kind),
        entry!(ForNode, "tirx.For", body),
        entry!(ForNode, "tirx.For", thread_binding),
        entry!(ForNode, "tirx.For", annotations),
        entry!(ForNode, "tirx.For", step),
        entry!(IfThenElseNode, "tirx.IfThenElse", condition),
        entry!(IfThenElseNode, "tirx.IfThenElse", then_case),
        entry!(IfThenElseNode, "tirx.IfThenElse", else_case),
        entry!(WhileNode, "tirx.While", condition),
        entry!(WhileNode, "tirx.While", body),
    ];

    let mut mismatches = Vec::new();
    for (rust, key, cxx, label) in table {
        match reflect_offset(key, cxx) {
            Ok(reflected) if reflected == rust => {}
            Ok(reflected) => mismatches.push(format!(
                "{label} @ {rust} (Rust) vs {key}.{cxx} @ {reflected} (C++)"
            )),
            Err(e) => mismatches.push(format!("{label}: {e}")),
        }
    }
    if mismatches.is_empty() {
        Ok(())
    } else {
        Err(format!(
            "tirx_ext: node layout mismatch against the loaded libtvm_compiler.so — \
             rebuild the extension against the current tir headers: {}",
            mismatches.join("; ")
        ))
    }
}
